File: work7/work7.cpp
#include <cmath>
#include <iostream>
#include <iomanip>

int main() {
    const double v0 = 17.0;
    const double k1 = 0.0089;
    const double g = 9.8;
    const double x = 3.0;
    const double y = 4.0;
    const double s = sqrt(x * x + y * y);
    const double z0 = 0.25;
    const int n = 15;

    double z_temp = z0;

    std::cout << std::fixed << std::setprecision(6);
    std::cout << "迭代补偿:\n";

    for (int i = 1; i <= n; ++i) {
        double theta = atan2(z_temp, s);
        double vx0 = v0 * cos(theta);
        double t = (exp(k1 * s) - 1) / (k1 * vx0);
        double z_real = v0 * sin(theta) * t - 0.5 * g * t * t;
        double dz = z0 - z_real;
        z_temp += dz;

        std::cout << "iter " << std::setw(2) << i
                  << "  theta =" << std::setw(10) << theta * 180.0 / M_PI
                  << "°  z_real =" << std::setw(10) << z_real
                  << "  dz =" << std::setw(10) << dz << std::endl;
    }

    double theta_final = atan2(z_temp, s);
    std::cout << "最终出射角: " << theta_final * 180.0 / M_PI << "°" << std::endl;

    return 0;
}

File: work7/work7.py
import math

v0 = 17.0
k1 = 0.0089
g = 9.8
x = 3.0
y = 4.0
s = math.sqrt(x * x + y * y)
z0 = 0.25

E = math.exp(k1 * s) - 1
A = E / k1
a = 0.5 * g * A * A / (v0 * v0)

D = A * A - 4 * a * (a + z0)

print(f"E = {E:.6f}, A = {A:.6f}, a = {a:.6f}, D = {D:.6f}\n")

if D < 0:
    print("判别式 D < 0，无实数解")
elif abs(D) < 1e-12:
    u = A / (2 * a)
    theta = math.degrees(math.atan(u))
    print(f"判别式 D ≈ 0，唯一解: θ = {theta:.6f}°")
else:
    u1 = (A + math.sqrt(D)) / (2 * a)
    u2 = (A - math.sqrt(D)) / (2 * a)
    theta1 = math.degrees(math.atan(u1))
    theta2 = math.degrees(math.atan(u2))

    print(f"高抛解 (u = {u1:.6f}): θ₁ = {theta1:.6f}°")
    print(f"低抛解 (u = {u2:.6f}): θ₂ = {theta2:.6f}°")
